Task 3

vershininm opened the pull request.

vershininm committed:
task-3: switch to github.com/stretchr/testify in tests (instead of own assertDeapEqual() func)

vershininm committed:
task-3: function that opens and closes file in a loop

vershininm committed:
task-3: introduce Request entity

vershininm committed:
task-3: helper functions to manipulate with slices of Request

applejohnny commented on internal/models/request.go:
```diff
@@ -0,0 +1,24 @@
+}
+
+// NewRequest create new Request instance
+func NewRequest(Id uint64, UserId uint64, Type uint64, Text string) *Request {
```
Структура `Request` очень простенькая и содержит просто набор полей, логичнее тут возвращать по значению, а не по ссылке.

applejohnny replied:
Аргументы функии не принято называть с большой буквы. 
Одинаковые типы можно сгруппировать вместе.

applejohnny commented on internal/models/request.go:
```diff
@@ -0,0 +1,24 @@
+
+// NewRequest create new Request instance
+func NewRequest(Id uint64, UserId uint64, Type uint64, Text string) *Request {
+	r := new(Request)
```
всё это можно сократить в запись `return Request{Id: Id,...}`

applejohnny commented on internal/models/request.go:
```diff
@@ -0,0 +1,24 @@
+
+import "fmt"
+
+type Request struct {
```
У всех экспортируемых типов должно быть описание, линтер будет ругаться :-)

applejohnny commented on internal/utils/files.go:
```diff
@@ -0,0 +1,27 @@
+				return
+			}
+			defer func() {
+				err = fp.Close()
```
В этом месте ты перетираешь ошибку открытия файла, т.е. файл не открылся, происходит возврат, вызывается `defer`, который затирает ошибку открытия.

applejohnny commented on internal/utils/requests.go:
```diff
@@ -0,0 +1,51 @@
+	"math"
+)
+
+type NonUniqueIdError struct {
```
Дока

applejohnny commented on internal/utils/requests.go:
```diff
@@ -0,0 +1,51 @@
+	DuplicateId uint64
+}
+
+func (err NonUniqueIdError) Error() string {
```
На экспортируемые методы тоже нужно писать доку.

applejohnny commented on internal/utils/requests.go:
```diff
@@ -0,0 +1,51 @@
+}
+
+// SplitToBulks Converts a given slice of Requests to a slice of slices of Requests of a given size.
+func SplitToBulks(items []*models.Request, chunkSize int) [][]*models.Request {
```
Лучше заменить ссылки на передачу по значению, т.е. `[]models.Request`, `[][]models.Request`

applejohnny commented on internal/utils/requests.go:
```diff
@@ -0,0 +1,51 @@
+
+// SplitToBulks Converts a given slice of Requests to a slice of slices of Requests of a given size.
+func SplitToBulks(items []*models.Request, chunkSize int) [][]*models.Request {
+	if chunkSize <= 0 {
```
Если ты в определении функции укажешь более точный тип `uint` на chunkSize, то эта проверка не понадобится.

applejohnny commented on internal/utils/requests.go:
```diff
@@ -0,0 +1,51 @@
+
+// SliceToMap Converts a given slice of Requests to a map where key is Request.Id and values are corresponding requests.
+// The function requires requests to have unique Id, otherwise it returns NonUniqueIdError.
+func SliceToMap(items []*models.Request) (map[uint64]*models.Request, error) {
```
Тоже самое, лучше `[]models.Request`, `map[uint64]models.Request`

applejohnny commented on internal/utils/requests_test.go:
```diff
@@ -0,0 +1,55 @@
+		[][]*models.Request{requests[:3], requests[3:]},
+		SplitToBulks(requests, 3),
+	)
+	assert.Equal(t, [][]*models.Request{}, SplitToBulks(requests, 0))
```
Раз уж у тебя `chunkSize int`, то стоит отдельно проверять на отрицательное значение. В случае `uint` этого делать уже не нужно будет.

applejohnny commented on internal/utils/requests_test.go:
```diff
@@ -0,0 +1,55 @@
+	byId, err := SliceToMap(requests)
+
+	assert.Equal(t, 3, len(byId))
+	assert.Equal(t, byId[1].Text, "request 1")
```
Пожалуй, тут логичнее будет сверять поле `Id`, т.к. оно подразумевается уникальным, а текст в разных обьектах может быть и одинаковым.

vershininm committed:
task-3: several minor fixes per review:
- pass/return Request by value (not by reference)
- add missing docs
- fix vars case in NewRequest()
- change chunkSize type from int to uint

vershininm committed:
task-3: add benchmark test for SplitToBulks

vershininm committed:
task-3: SplitToBulksInt() chunkSize param switched to uint

vershininm committed:
task-3: add missing docstring to Request.String()
